### PR TITLE
Experiment impact visualization

### PR DESCRIPTION
**Pull request by mullr**

### CODE REVIEW

File: vscode/resources/transitionGraph.js [hunk 1]
```diff
@@ -62,42 +82,36 @@
...
                     y: ele.position("y"),
                 };
             }
         });
+        savePersistentState();
     }
 
     function savePersistentState() {
-        vscode.setState({ nodeElements, edgeElements, nodeCoordinates, layout: selectedLayout });
+        vscode.setState(persistentState);
```
Comment (jonlamb-gh) [inline vscode/resources/transitionGraph.js:100]: I was under the impression that we don't need to use the getState/setState APIs when the `retainContextWhenHidden` webview option is enabled.

File: vscode/src/experiments.ts [hunk 2]
```diff
@@ -69,6 +79,127 @@ export class ExperimentsTreeDataProvider implements vscode.TreeDataProvider<Expe
...
+            assignNodeProps.addClass(impact.timelineName, "impact");
+            assignNodeProps.addDataProp(impact.timelineName, "severity", impact.severity);
+            assignNodeProps.addDataProp(impact.timelineName, "impactHtml", impact.detailsHtml);
+        }
+
+        vscode.commands.executeCommand("auxon.transition.graph", {
+            type: "segment",
+            segmentIds,
+            title: `Experiment Impact for scenario '${scenario.scenarioName}'`,
+            //title: `Experiment Impact for '${scenario.experimentName}' scenario '${scenario.scenarioName}'`,
```
Comment (jonlamb-gh) [inline vscode/src/experiments.ts:197]: Can this be removed?

File: vscode/src/transitionGraph.ts [hunk 3]
```diff
@@ -313,7 +325,18 @@ export class TransitionGraph {
         if (params.type == "timelines") {
             res = await this.apiClient.timelines().groupedGraph(params.timelines, params.groupBy);
         } else if (params.type == "segment") {
-            res = await this.apiClient.segment(params.segmentId).groupedGraph(params.groupBy);
+            if (params.segmentIds.length == 1) {
+                res = await this.apiClient.segment(params.segmentIds[0]).groupedGraph(params.groupBy);
+            } else {
+                const timelineIds: api.TimelineId[] = [];
+                // TODO Not ideal...
```
Comment (jonlamb-gh) [inline vscode/src/transitionGraph.ts:332]: Should we make this better now?

File: vscode/src/transitionGraph.ts [hunk 4]
```diff
@@ -396,10 +436,35 @@ class DirectedGraph {
...
+    timelineName?: string;
+    eventName?: string;
+    severity?: number;
+    impactHtml?: string;
+    count?: number;
+}
 
 class Node {
     description?: string = undefined;
+    // TODO unused? remove?
```
Comment (jonlamb-gh) [inline vscode/src/transitionGraph.ts:467]: I don't think it's currently used. I had planned on exposing source file/line URIs for click-to-open-in-editor, but haven't plumbed that yet.

File: vscode/src/transitionGraph.ts [hunk 5]
```diff
@@ -408,92 +473,130 @@ class Node {
...
+    //     if (this.eventName !== undefined && this.timelineId !== undefined) {
+    //         return {
+    //             name: this.eventName,
+    //             timeline: { id: this.timelineId, name: this.timelineName },
+    //             count: this.count,
+    //         };
+    //     } else {
+    //         return undefined;
+    //     }
+    // }
```
Comment (jonlamb-gh) [inline vscode/src/transitionGraph.ts:556]: This can be removed now, right?

File: vscode/templates/transitionGraph.html [hunk 6]
```diff
@@ -26,35 +26,63 @@
...
                 </vscode-button>
-                <vscode-button id="toolbarRefresh" appearance="icon" aria-label="Refresh">
-                    <span class="codicon codicon-refresh"></span>
+                <vscode-button id="toolbarRefresh" appearance="secondary" aria-label="Refresh Layout">
+                    Refresh Layout
+                    <span slot="start" class="codicon codicon-refresh"></span>
                 </vscode-button>
+                <!-- <vscode-button id="toolbarCollapse" appearance="icon" aria-label="Collapse Sidebar"> -->
+                <!--     <span class="codicon codicon-chevron-down"></span> -->
+                <!-- </vscode-button> -->
```
Comment (jonlamb-gh) [inline vscode/templates/transitionGraph.html:43]: This can be removed now too?

**Review by jonlamb-gh — APPROVED** [review]

File: vscode/package.json [hunk 7]
```diff
@@ -897,12 +888,16 @@
...
+                "title": "Show Experiment Impact Report",
+                "icon": "$(pulse)"
+            },
+            {
+                "command": "auxon.experiments.visualizeImpactScenario",
+                "title": "Visual Experiment Impact Scenario"
             }
         ],
         "configuration": {
-            "type": "object",
```
[the remark this answers is not in the document]
Comment (mullr) [reply vscode/package.json:905]: nope

File: vscode/src/transitionGraph.ts (hunk 3 above)
Comment (mullr) [reply vscode/src/transitionGraph.ts:332]: I think it's fine for now, but I ticketed it.  #2714

File: vscode/resources/transitionGraph.js (hunk 1 above)
Comment (mullr) [reply vscode/resources/transitionGraph.js:100]: I think you're right, but that option is not enabled for the transition graph.

File: vscode/src/transitionGraph.ts (hunk 4 above)
Comment (mullr) [reply vscode/src/transitionGraph.ts:467]: Okay, I'll remove it for now just for cleanliness, will be easy to add back.

**Pull request merged**